src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -921,59 +927,55 @@ private string ConsumeTokenToString(int characterCount)
         /// all indeces to the characterBuffer are invalid except for tokenStartIndex.</remarks>
         private bool ReadInput()
         {
-            Debug.Assert(this.storedCharacterCount <= this.characterBuffer.Length, "We can only store as many characters as fit into our buffer.");
             Debug.Assert(this.tokenStartIndex >= 0 && this.tokenStartIndex <= this.storedCharacterCount, "The token start is out of stored characters range.");
 
             if (this.endOfInputReached)
             {
                 return false;
             }
 
+            // initialze the buffer
+            if (this.characterBuffer == null)
+            {
+                this.characterBuffer = BufferUtils.RentFromBuffer(ArrayPool, InitialCharacterBufferSize);
+            }
+            Debug.Assert(this.storedCharacterCount <= this.characterBuffer.Length, "We can only store as many characters as fit into our buffer.");
+
             // We need to make sure we have more room in the buffer to read characters into
             if (this.storedCharacterCount == this.characterBuffer.Length)
             {
                 // No more room in the buffer, move or grow the buffer.
                 if (this.tokenStartIndex == this.storedCharacterCount)
                 {
-                    // If the buffer is empty (all characters were consumed from it)
-                    // just start over.
+                    // If the buffer is empty (all characters were consumed from it), just start over.
                     this.tokenStartIndex = 0;
                     this.storedCharacterCount = 0;
                 }
-                else if (this.tokenStartIndex > (this.characterBuffer.Length - MaxCharacterCountToMove))
-                {
-                    // Some characters were consumed, we can just move them in the buffer
-                    // to get more room without allocating.
-                    Array.Copy(
-                        this.characterBuffer,
-                        this.tokenStartIndex,
-                        this.characterBuffer,
-                        0,
-                        this.storedCharacterCount - this.tokenStartIndex);
-                    this.storedCharacterCount -= this.tokenStartIndex;
-                    this.tokenStartIndex = 0;
-                }
-                else
+                else if (this.tokenStartIndex < this.characterBuffer.Length / 2)
                 {
                     // The entire buffer is full of unconsumed characters
-                    // We need to grow the buffer.
-                    // Double the size of the buffer.
+                    // We need to grow the buffer. Double the size of the buffer.
                     int newBufferSize = this.characterBuffer.Length * 2;
-                    char[] newCharacterBuffer = new char[newBufferSize];
+                    newBufferSize = newBufferSize < 0 ? int.MaxValue : newBufferSize; // maybe overflow

```

Comment:
What happens if we reach int.MaxValue, and decide that we need more memory?

Reply:
I add a check before double the buffer. 

Please take a look if that's ok?

---
In reply to: [253993003](https://github.com/OData/odata.net/pull/1388#discussion_r253993003) [](ancestors = 253993003)